# Validation run analysis - Normal behaviour
number_of_stages = 1 
number_of_visitors = 100 
number_of_crew = 0 # dont need crew for normal behaviour
vision_visitor = 15  
vision_crew = 8 
cell_capacity = 16 
max_group_size = 5 
alarm_timer = 300 # only testing normal behaviour - alarm should not go off
number_of_steps = 300 
number_of_runs = 1 

In [70]:
import pandas as pd

In [71]:
agent_data = pd.read_csv('validation_run.csv')
agent_data

Step  AgentID                               State  Toilet timer  \
0         0        1                                 NaN          2477   
1         0        2                                 NaN          1930   
2         0        3                                 NaN           990   
3         0        4                                 NaN           661   
4         0        5                                 NaN          1350   
...     ...      ...                                 ...           ...   
30095   300       96  Friends close by - moving randomly           531   
30096   300       97  Friends close by - moving randomly          1826   
30097   300       98  Friends close by - moving randomly          -260   
30098   300       99  Friends close by - moving randomly           324   
30099   300      100  Friends close by - moving randomly          1171   

       Toilet delay  Bar timer  Bar delay Current Location Current floor  \
0               245       1379        143         (81, 35)           NaN   
1               142        618        152         (85, 34)           NaN   
2                40        505         71         (81, 19)           NaN   
3               123        897        140         (83, 22)           NaN   
4               256         20        115         (82, 19)           NaN   
...             ...        ...        ...              ...           ...   
30095            34        181        134         (46, 19)    ['ground']   
30096           115       1039         87         (16, 12)    ['ground']   
30097            68         21        107         (11, 18)    ['ground']   
30098           236        825         63         (49, 22)    ['ground']   
30099           237        403        109         (53, 19)    ['ground']   

      Target Location Target floor Stair target {Previous location  \
0                 NaN          NaN          NaN                NaN   
1                 NaN          NaN          NaN                NaN   
2                 NaN          NaN          NaN                NaN   
3                 NaN          NaN          NaN                NaN   
4                 NaN          NaN          NaN                NaN   
...               ...          ...          ...                ...   
30095             NaN   ['ground']          NaN           (46, 19)   
30096             NaN   ['ground']          NaN           (16, 12)   
30097             NaN   ['ground']          NaN           (11, 18)   
30098             NaN   ['ground']          NaN           (49, 22)   
30099             NaN   ['ground']          NaN           (53, 19)   

            Friends                               Agents in vision far  \
0         [2, 3, 4]                                                NaN   
1         [1, 3, 4]                                                NaN   
2         [1, 2, 4]                                                NaN   
3         [1, 2, 3]                                                NaN   
4            [6, 7]                                                NaN   
...             ...                                                ...   
30095  [95, 97, 98]  [84, 70, 60, 83, 86, 63, 56, 91, 42, 87, 54, 5...   
30096  [95, 96, 98]  [18, 43, 79, 55, 59, 49, 3, 39, 20, 16, 28, 41...   
30097  [95, 96, 97]  [18, 43, 79, 55, 59, 49, 3, 39, 5, 20, 16, 28,...   
30098         [100]  [70, 72, 60, 83, 86, 74, 63, 56, 91, 42, 87, 5...   
30099          [99]  [83, 86, 63, 56, 91, 42, 87, 54, 52, 36, 66, 9...   

      Friends in vision far  \
0                       NaN   
1                       NaN   
2                       NaN   
3                       NaN   
4                       NaN   
...                     ...   
30095                  [95]   
30096                  [98]   
30097                  [97]   
30098                 [100]   
30099                  [99]   

                                  Agents in vision close  \
0                                         

First check the state variable - this is the first easy check to see if all parts of the code have been reached and thus executed. The agents in the normal behaviour should have 11 states (12 including the None from the initialisation)

In [72]:
agent_data['State'].unique()

array([nan, 'Friends close by - moving randomly',
       'No friends in vision - moving randomly',
       'Friends far away - moving closer',
       'Friends far away - not visible', 'Moving towards toilet',
       'Moving towards bar',
       'Friends far away - timer went off - moving randomly',
       'Waiting at bar', 'Moving towards previous location',
       'Arrived at previous location', 'Waiting at toilet'], dtype=object)

All states are present in the validation run. Now lets look at some specific examples to see if the code is implemented correctly:

## Timer logic

In [73]:
# Look at a random agent
agent_data[agent_data['AgentID']==79]

Step  AgentID                               State  Toilet timer  \
78        0       79                                 NaN           167   
178       1       79  Friends close by - moving randomly           166   
278       2       79  Friends close by - moving randomly           165   
378       3       79  Friends close by - moving randomly           164   
478       4       79  Friends close by - moving randomly           163   
...     ...      ...                                 ...           ...   
29678   296       79                   Waiting at toilet             0   
29778   297       79                   Waiting at toilet             0   
29878   298       79                   Waiting at toilet             0   
29978   299       79                   Waiting at toilet             0   
30078   300       79                   Waiting at toilet             0   

       Toilet delay  Bar timer  Bar delay Current Location Current floor  \
78              280       1613         75         (50, 24)           NaN   
178             280       1612         75         (50, 25)           NaN   
278             280       1611         75         (50, 26)           NaN   
378             280       1610         75         (51, 26)           NaN   
478             280       1609         75         (50, 25)           NaN   
...             ...        ...        ...              ...           ...   
29678           191       1446         75           (4, 7)    ['ground']   
29778           190       1446         75           (4, 7)    ['ground']   
29878           189       1446         75           (4, 7)    ['ground']   
29978           188       1446         75           (4, 7)    ['ground']   
30078           187       1446         75           (4, 7)    ['ground']   

      Target Location Target floor Stair target {Previous location  \
78                NaN          NaN          NaN                NaN   
178               NaN          NaN          NaN           (50, 25)   
278               NaN          NaN          NaN           (50, 26)   
378               NaN          NaN          NaN           (51, 26)   
478               NaN          NaN          NaN           (50, 25)   
...               ...          ...          ...                ...   
29678          (4, 7)   ['ground']          NaN           (40, 18)   
29778          (4, 7)   ['ground']          NaN           (40, 18)   
29878          (4, 7)   ['ground']          NaN           (40, 18)   
29978          (4, 7)   ['ground']          NaN           (40, 18)   
30078          (4, 7)   ['ground']          NaN           (40, 18)   

                Friends                               Agents in vision far  \
78     [78, 80, 81, 82]                                                NaN   
178    [78, 80, 81, 82]  [40, 99, 86, 57, 75, 87, 100, 61, 70, 51, 80, ...   
278    [78, 80, 81, 82]  [40, 75, 99, 87, 100, 61, 86, 57, 85, 98, 70, ...   
378    [78, 80, 81, 82]  [40, 99, 75, 87, 61, 86, 85, 70, 100, 98, 65, ...   
478    [78, 80, 81, 82]  [99, 40, 86, 70, 75, 87, 61, 51, 59, 85, 98, 9...   
...                 ...                                                ...   
29678  [78, 80, 81, 82]  [96, 71, 92, 86, 74, 75, 81, 95, 76, 83, 56, 5...   
29778  [78, 80, 81, 82]  [96, 71, 92, 86, 74, 75, 81, 95, 76, 83, 56, 5...   
29878  [78, 80, 81, 82]  [96, 71, 92, 86, 74, 75, 81, 95, 76, 83, 56, 5...   
29978  [78, 80, 81, 82]  [96, 71, 92, 86, 74, 75, 81, 95, 76, 83, 56, 5...   
30078  [78, 80, 81, 82]  [96, 71, 92, 86, 74, 75, 81, 95, 76, 83, 56, 5...   

      Friends in vision far  \
78                      NaN   
178        [78, 80, 81, 82]   
278        [78, 80, 81, 82]   
378        [78, 80, 81, 82]   
478        [78, 80, 81, 82]   
...                     ...   
29678      [78, 80, 81, 82]   
29778      [78, 80, 81, 82]   
29878      [78, 80, 81, 82]   
29978      [78, 80, 81, 82]   
30078      [78, 80, 81, 82]   

                                  Agents in vision clo

As can be seen in the dataframe above, the timers go down by one each time step, just like they are supposed to. As per the last else part of the normal_behaviour function; the countdown of timer A should freeze the moment timer B goes off vise versa. In the data above, it becomes clear that this works around timestep 167. Looking at the next step - the moment the toilet timer of the agent goes off, it stops looking at its friends and starts moving towards the toilet. The target location is set on the locations of the toilets in the map, and it is visible that each time step, the agents moves closer towards its goal. It is also visible that the agents saves its previous location/ current location each time step. The moment the timer goes off, the previous location does not get updated anymore. This way the agent can return to the location it was whenever it is done doing its tasks. At timestep 208 the agent reaches the toilets. This can be seen by the current location that matches the target location, as well as the shift in state (from moving towards the toilet to waiting at toilet). From that point on the toilet delay also starts counting down, indicating that the code works. As the validation run was only done with 300 time steps, this won't be shown anymore.

In [74]:
# Look at another random agent
agent_data[agent_data['AgentID']==54]

Step  AgentID                                   State  Toilet timer  \
53        0       54                                     NaN           256   
153       1       54      Friends close by - moving randomly           255   
253       2       54      Friends close by - moving randomly           254   
353       3       54      Friends close by - moving randomly           253   
453       4       54          Friends far away - not visible           252   
...     ...      ...                                     ...           ...   
29653   296       54  No friends in vision - moving randomly           188   
29753   297       54  No friends in vision - moving randomly           188   
29853   298       54                      Moving towards bar           188   
29953   299       54                      Moving towards bar           188   
30053   300       54                      Moving towards bar           188   

       Toilet delay  Bar timer  Bar delay Current Location Current floor  \
53              204          7        136         (45, 25)           NaN   
153             204          6        136         (46, 25)           NaN   
253             204          5        136         (47, 26)           NaN   
353             204          4        136         (46, 26)           NaN   
453             204          3        136         (45, 26)           NaN   
...             ...        ...        ...              ...           ...   
29653           204          1          9         (45, 15)    ['ground']   
29753           204          0          9         (45, 16)    ['ground']   
29853           204          0          9         (45, 16)    ['ground']   
29953           204          0          9         (44, 17)    ['ground']   
30053           204          0          9         (43, 18)    ['ground']   

      Target Location Target floor Stair target {Previous location Friends  \
53                NaN          NaN          NaN                NaN    [55]   
153               NaN          NaN          NaN           (46, 25)    [55]   
253               NaN          NaN          NaN           (47, 26)    [55]   
353               NaN          NaN          NaN           (46, 26)    [55]   
453               NaN          NaN          NaN           (45, 26)    [55]   
...               ...          ...          ...                ...     ...   
29653             NaN   ['ground']          NaN           (45, 15)    [55]   
29753             NaN   ['ground']          NaN           (45, 16)    [55]   
29853        (42, 28)   ['ground']          NaN           (45, 16)    [55]   
29953        (42, 28)   ['ground']          NaN           (45, 16)    [55]   
30053        (42, 28)   ['ground']          NaN           (45, 16)    [55]   

                                    Agents in vision far  \
53                                                   NaN   
153    [40, 99, 70, 86, 87, 100, 77, 61, 51, 57, 80, ...   
253    [40, 99, 57, 75, 87, 100, 61, 70, 86, 85, 98, ...   
353    [40, 99, 75, 87, 86, 85, 61, 57, 100, 98, 70, ...   
453    [99, 40, 86, 75, 61, 87, 70, 51, 85, 98, 65, 5...   
...                                                  ...   
29653  [84, 70, 86, 60, 83, 63, 42, 87, 66, 85, 46, 9...   
29753  [84, 70, 60, 86, 83, 66, 63, 42, 87, 36, 91, 5...   
29853  [84, 70, 60, 86, 83, 66, 63, 42, 87, 36, 91, 5...   
29953  [84, 70, 60, 86, 83, 66, 63, 42, 87, 36, 91, 5...   
30053  [84, 70, 60, 86, 83, 66, 63, 42, 87, 36, 91, 5...   

      Friends in vision far  \
53                      NaN   
153                    [55]   
253                    [55]   
353                    [55]   
453                    [55]   
...                     ...   
29653                    []   
29753                    []   
29853                    []   
29953                    []   
30053                    []   

                                  Agents in vision close  \
53                                                   NaN   
153   

That the code for returning to the previous location works can be seen when looking at agent 54. At step 7 its bar timer goes of, and since it is already standing close to the bar, it reaches the bar at step 14. The agent has a long bar delay, at time step 150 this delay is done. As can be seen by the change in state as well as the target location, it shows that the agents moves back to where it came from. The moment the agent reaches its old location (where it was before the bar timer went off), the target location resets and it starts looking for friends.

## Find friends function
Agent 54 can also be used to show the workings of the find friends function. The moment the agent returns to the location it was when its bar timer went off (Setp 154), it starts looking for its friends. This agent has only one friend - agent 55. Looking at the columns for friends in vision far and close, it becomes clear that the friend is visible in the far away vision, but not in the close one. The columns for Agents in vision far and close serve as an additional check to see if the code works - agent 55 is part of all of the agents agent 54 can see in its far away vision, but not in the close by vision. This means that agent 54 has to move closer towards agent 55 as to not loose them. Looking at the state variable, this also happens; agent 55 notices that its friend far away, so it is moving closer. Twice the state 'Friends far away - not visible' is shown. This can be the case because agent 54 (the friend) is also moving. If it happens to move behind an obstacle, agent 55 cannot see it, and thus cannot move towards it (it will move randomly). Luckily, the friend are reunited shortly afterwards in step 159. The agents are close enough together that they are moving randomly (or dancing as you will) instead of towards each other. This distance between the friends is evaluated each time step to make sure that they dont move to far away from each other and get left alone. 

## Stair target

In [75]:
# Random agent that takes the stairs
agent_data[agent_data['AgentID']==11]

Step  AgentID                               State  Toilet timer  \
10        0       11                                 NaN            40   
110       1       11  Friends close by - moving randomly            39   
210       2       11  Friends close by - moving randomly            38   
310       3       11  Friends close by - moving randomly            37   
410       4       11  Friends close by - moving randomly            36   
...     ...      ...                                 ...           ...   
29610   296       11  Friends close by - moving randomly          -236   
29710   297       11  Friends close by - moving randomly          -237   
29810   298       11  Friends close by - moving randomly          -238   
29910   299       11  Friends close by - moving randomly          -239   
30010   300       11  Friends close by - moving randomly          -240   

       Toilet delay  Bar timer  Bar delay Current Location Current floor  \
10              269       1231        215         (79, 12)           NaN   
110             269       1230        215         (79, 11)           NaN   
210             269       1229        215         (79, 10)           NaN   
310             269       1228        215         (79, 11)           NaN   
410             269       1227        215         (79, 10)           NaN   
...             ...        ...        ...              ...           ...   
29610           269       1191        215         (13, 11)    ['ground']   
29710           269       1191        215         (12, 10)    ['ground']   
29810           269       1191        215         (13, 11)    ['ground']   
29910           269       1191        215         (12, 10)    ['ground']   
30010           269       1191        215         (13, 11)    ['ground']   

      Target Location Target floor Stair target {Previous location  \
10                NaN          NaN          NaN                NaN   
110               NaN          NaN          NaN           (79, 11)   
210               NaN          NaN          NaN           (79, 10)   
310               NaN          NaN          NaN           (79, 11)   
410               NaN          NaN          NaN           (79, 10)   
...               ...          ...          ...                ...   
29610             NaN   ['ground']          NaN           (13, 11)   
29710             NaN   ['ground']          NaN           (12, 10)   
29810             NaN   ['ground']          NaN           (13, 11)   
29910             NaN   ['ground']          NaN           (12, 10)   
30010             NaN   ['ground']          NaN           (13, 11)   

                Friends                               Agents in vision far  \
10     [10, 12, 13, 14]                                                NaN   
110    [10, 12, 13, 14]  [6, 22, 10, 12, 3, 30, 16, 14, 24, 23, 21, 15,...   
210    [10, 12, 13, 14]  [12, 22, 10, 6, 3, 16, 30, 24, 29, 23, 15, 14,...   
310    [10, 12, 13, 14]  [22, 3, 12, 10, 6, 30, 16, 29, 23, 24, 5, 4, 1...   
410    [10, 12, 13, 14]  [22, 12, 10, 6, 3, 30, 23, 29, 14, 16, 24, 5, ...   
...                 ...                                                ...   
29610  [10, 12, 13, 14]  [3, 18, 79, 55, 59, 39, 49, 43, 41, 20, 24, 37...   
29710  [10, 12, 13, 14]  [3, 79, 55, 59, 18, 39, 49, 43, 12, 41, 24, 14...   
29810  [10, 12, 13, 14]  [3, 79, 55, 59, 18, 49, 39, 43, 14, 24, 41, 35...   
29910  [10, 12, 13, 14]  [18, 3, 79, 55, 59, 49, 39, 43, 12, 16, 28, 41...   
30010  [10, 12, 13, 14]  [39, 79, 55, 59, 18, 3, 43, 49, 16, 28, 41, 35...   

      Friends in vision far  \
10                      NaN   
110        [10, 12, 14, 13]   
210        [10, 12, 14, 13]   
310        [10, 12, 14, 13]   
410        [10, 12, 14, 13]   
...                     ...   
29610                  [10]   
29710          [10, 12, 14]   
29810              [10, 14]   
29910              [10, 12]   
30010                  [10]   

                                  Agents in vision clo

The agent above shows the workings of the take stars function. Around step 40, the agents toilet timer goes off. The moment this happens, the code checks at what floor the agent currently is, and on what floor their target location is. In this case, the location of the toilet is on the ground floor, while the agent is at the first floor. The agent thus can't directly move towards its target (as it will get stuck walking against a wall), but has to take the stair first. On step 41, this check is done, and an additional target location (Stair target) is added to the agent. The agent first walks towards the stairs - as can be seen by following its current location. When it has reached the stair target, it is transported to the fround floor (step 47), the stair target disappears, and the agent can continue its way towards the toilet. 

# Validation run - evacuation behaviour
number_of_stages = 1 
number_of_visitors = 100 
number_of_crew = 7
vision_visitor = 15  
vision_crew = 8 
cell_capacity = 16 
max_group_size = 5 
alarm_timer = 5 # Few steps of normal behaviour to show that the timer works
number_of_steps = 300 
number_of_runs = 1 

In [76]:
agent_data_evac = pd.read_csv('validation_run_evac.csv')
agent_data_evac

Step  AgentID Found Exit State Current Location Target Location  \
0        0        1      False   NaN         (77, 14)             NaN   
1        0        2      False   NaN          (80, 6)             NaN   
2        0        3      False   NaN         (85, 18)             NaN   
3        0        4      False   NaN         (79, 16)             NaN   
4        0        5      False   NaN         (79, 24)             NaN   
...    ...      ...        ...   ...              ...             ...   
5430   300      103        NaN   NaN              NaN             NaN   
5431   300      104        NaN   NaN              NaN             NaN   
5432   300      105        NaN   NaN              NaN             NaN   
5433   300      106        NaN   NaN              NaN             NaN   
5434   300      107        NaN   NaN              NaN             NaN   

        Friends Friends in vision  Search timer Seen crew Social Observer  
0     [2, 3, 4]               NaN          23.0     False   True    False  
1     [1, 3, 4]               NaN          16.0     False  False     True  
2     [1, 2, 4]               NaN          10.0     False   True    False  
3     [1, 2, 3]               NaN          28.0     False  False    False  
4        [6, 7]               NaN          14.0     False   True    False  
...         ...               ...           ...       ...    ...      ...  
5430        NaN               NaN           NaN       NaN    NaN      NaN  
5431        NaN               NaN           NaN       NaN    NaN      NaN  
5432        NaN               NaN           NaN       NaN    NaN      NaN  
5433        NaN               NaN           NaN       NaN    NaN      NaN  
5434        NaN               NaN           NaN       NaN    NaN      NaN  

[5435 rows x 12 columns]

The first check is again looking at the states. There are 8 evacuation states. As the first few steps are still the normal behaviour, there will be more states than that.

In [77]:
agent_data_evac['State'].unique()

array([nan, 'Friends close by - moving randomly',
       'No friends in vision - moving randomly',
       'Friends far away - not visible',
       'Friends far away - moving closer', 'Moving towards toilet',
       'Friend found with NO target location - go to main exit',
       'Found close exit - moving towards',
       'Friend found with target location - stolen target',
       'IDGAF - moving towards main entrance',
       'Friend found with NO target location - observing exits',
       "Hasn't found close exit - moving randomly",
       'No friends in vision - moving randomly till timer',
       'No friends in vision - moving randomly till timer or see friends',
       'Seen crew - moving towards nearest exit'], dtype=object)

The only evacuation state that is missing in this run is the 'Friend found with NO target location - observing exits'. However, as this is very well possible, the assumption is made that the code works regardless. 

## Search timer & observer

In [78]:
agent_data_evac[agent_data_evac['AgentID']==57]

Step  AgentID Found Exit  \
56       0       57      False   
163      1       57      False   
270      2       57      False   
377      3       57      False   
484      4       57      False   
591      5       57      False   
698      6       57      False   
804      7       57      False   
909      8       57      False   
1014     9       57      False   
1118    10       57      False   
1218    11       57      False   
1316    12       57      False   
1412    13       57      False   
1502    14       57      False   
1589    15       57      False   
1673    16       57      False   
1754    17       57      False   
1832    18       57      False   
1908    19       57      False   
1980    20       57      False   
2052    21       57      False   
2123    22       57      False   
2192    23       57      False   
2261    24       57      False   
2328    25       57      False   
2394    26       57      False   
2456    27       57      False   
2516    28       57      False   
2575    29       57      False   
2633    30       57      False   
2688    31       57      False   
2739    32       57      False   
2787    33       57      False   
2834    34       57      False   
2880    35       57      False   

                                                  State Current Location  \
56                                                  NaN         (47, 16)   
163              No friends in vision - moving randomly         (46, 16)   
270              No friends in vision - moving randomly         (47, 17)   
377              No friends in vision - moving randomly         (48, 16)   
484              No friends in vision - moving randomly         (48, 17)   
591              No friends in vision - moving randomly         (48, 18)   
698   No friends in vision - moving randomly till ti...         (47, 18)   
804   No friends in vision - moving randomly till ti...         (46, 17)   
909   No friends in vision - moving randomly till ti...         (46, 16)   
1014  No friends in vision - moving randomly till ti...         (46, 15)   
1118  No friends in vision - moving randomly till ti...         (45, 16)   
1218  No friends in vision - moving randomly till ti...         (45, 17)   
1316  No friends in vision - moving randomly till ti...         (44, 17)   
1412  No friends in vision - moving randomly till ti...         (45, 18)   
1502  No friends in vision - moving randomly till ti...         (45, 17)   
1589  No friends in vision - moving randomly till ti...         (45, 18)   
1673  No friends in vision - moving randomly till ti...         (44, 18)   
1754  No friends in vision - moving randomly till ti...         (44, 19)   
1832  No friends in vision - moving randomly till ti...         (44, 18)   
1908  No friends in vision - moving randomly till ti...         (44, 17)   
1980  No friends in vision - moving randomly till ti...         (45, 17)   
2052                  Found close exit - moving towards         (46, 16)   
2123                  Found close exit - moving towards         (47, 15)   
2192                  Found close exit - moving towards         (48, 15)   
2261                  Found close exit - moving towards         (49, 15)   
2328                  Found close exit - moving towards         (50, 15)   
2394            Seen crew - moving towards nearest exit         (51, 15)   
2456            Seen crew - moving towards nearest exit         (52, 15)   
2516            Seen crew - moving towards nearest exit         (53, 14)   
2575            Seen crew - moving towards nearest exit         (54, 13)   
2633            Seen crew - moving towards nearest exit         (55, 12)   
2688            Seen crew - moving towards nearest exit         (56, 11)   
2739            Seen crew - moving towards nearest exit         (57, 10)   
2787            Seen crew - moving towards nearest exit          (58, 9)   
2834            Seen crew - moving towards nearest exit          (59, 8)   
2880       

Social agents will first try and find their friends, before evacuating. However, if after a certain time (search timer) they havnt found any of their friends, they will move towards an exit. Depending on if the social agents is an observer or not, they will either look for the closest exit, or move towards a known exit (which is the main entrance). Agent 57 above is a social observer. As can be seen in the dataframe, the agent moves randomly trying to find their friend. Around step 20, when the search timer has gone off, the agent still hasnt found their friend. As this agent is an observer, it starts looking around for an exit while moving randomly. They find one at once, and move towards it. However, when later it spots a crew member, it adapts their target location to the closer exit (which they have been made aware of by the crew, for more see below).

## Seen crew

In [97]:
agent_data_evac[agent_data_evac['AgentID']==7]

Step  AgentID Found Exit  \
6        0        7      False   
113      1        7      False   
220      2        7      False   
327      3        7      False   
434      4        7      False   
541      5        7      False   
648      6        7      False   
755      7        7      False   
861      8        7      False   
966      9        7      False   
1071    10        7      False   
1173    11        7      False   
1273    12        7      False   
1371    13        7      False   
1466    14        7      False   
1556    15        7      False   
1641    16        7      False   
1722    17        7      False   
1800    18        7      False   
1876    19        7      False   
1951    20        7      False   
2023    21        7      False   
2094    22        7      False   
2164    23        7      False   
2233    24        7      False   
2301    25        7      False   
2367    26        7      False   
2430    27        7      False   
2492    28        7      False   
2552    29        7      False   
2611    30        7      False   

                                                  State Current Location  \
6                                                   NaN         (84, 14)   
113                  Friends close by - moving randomly         (83, 14)   
220                  Friends close by - moving randomly         (82, 14)   
327                  Friends close by - moving randomly         (83, 14)   
434                  Friends close by - moving randomly         (83, 13)   
541                  Friends close by - moving randomly         (84, 13)   
648   Friend found with target location - stolen target         (83, 12)   
755             Seen crew - moving towards nearest exit         (82, 11)   
861             Seen crew - moving towards nearest exit         (81, 10)   
966             Seen crew - moving towards nearest exit          (80, 9)   
1071            Seen crew - moving towards nearest exit          (79, 8)   
1173            Seen crew - moving towards nearest exit          (78, 7)   
1273            Seen crew - moving towards nearest exit          (77, 8)   
1371            Seen crew - moving towards nearest exit          (76, 8)   
1466            Seen crew - moving towards nearest exit          (75, 8)   
1556            Seen crew - moving towards nearest exit          (74, 8)   
1641            Seen crew - moving towards nearest exit          (73, 8)   
1722            Seen crew - moving towards nearest exit          (72, 9)   
1800            Seen crew - moving towards nearest exit          (9, 12)   
1876            Seen crew - moving towards nearest exit         (10, 12)   
1951            Seen crew - moving towards nearest exit         (11, 13)   
2023            Seen crew - moving towards nearest exit         (12, 14)   
2094            Seen crew - moving towards nearest exit         (13, 13)   
2164            Seen crew - moving towards nearest exit         (14, 12)   
2233            Seen crew - moving towards nearest exit         (15, 11)   
2301            Seen crew - moving towards nearest exit         (16, 10)   
2367            Seen crew - moving towards nearest exit          (17, 9)   
2430            Seen crew - moving towards nearest exit          (18, 8)   
2492            Seen crew - moving towards nearest exit          (19, 7)   
2552            Seen crew - moving towards nearest exit          (20, 6)   
2611            Seen crew - moving towards nearest exit          (21, 6)   

     Target Location Friends Friends in vision  Search timer Seen crew Social  \
6                NaN  [5, 6]               NaN          12.0     False   True   
113              NaN  [5, 6]               NaN          12.0     False   True   
220              NaN  [5, 6]               NaN          12.0     False   True   
327              NaN  [5, 6]               NaN          12.0     False   True   
434              NaN  [5, 6]               NaN          12.0     False   True  

Agent 7 above shows the importance of the crew even more. This agent is with their friend the moment the alarm goes of. Both of its friends are socials but not observers (see below, only agent 5 is visible for agent 7, but the agent reporter was not implemented correctly so the column does not give a value), meaning that the moment they have spotted a friend, they will walk towards a known exit together, which is the main entrance. This is also visible in the current location column; the moment the agent has found a friend, it takes over their target location and moves towards it.  (As is inherent with ABM models, there is an order of education. In this case, agent 5 does all of their steps before agent 7. Agent 5 has found a friend, and as they are not an observer, sets their target location to the main entrance. Agent 7 does their steps afterwards - they have found their friend agent 5, which has a target location. Agent 7 takes over this target location.). Immediately after, they run into a crew member, which points them towards a closer exit (see the change in state and target location), ensuring that the agents can evacuate as fast and safe as possible. 